sphinxlint.py

```diff
@@ -196,124 +197,206 @@ def is_in_a_table(error, line):
+@checker(".rst")
+def check_directives_using_three_dots(file, lines):
+    """Search for directives with three dots instead of two."""
     for lno, line in enumerate(lines, start=1):
-        if seems_directive_re.search(line):
-            yield lno, "comment seems to be intended as a directive"
         if three_dot_directive_re.search(line):
             yield lno, "directive should start with two dots, not three."
 
 
-@checker(".rst", severity=2)
-def check_role_missing_surrogate_escape(file, lines):
-    # Find role glued with a plural mark or something like:
-    #    The :exc:`Exception`s
-    # instead of:
-    #    The :exc:`Exceptions`\ s
-    # The difficulty here is that the following is valid:
-    #    The :literal:`:exc:`Exceptions``
-    # While this is not:
-    #    The :literal:`:exc:`Exceptions``s
+@checker(".rst")
+def check_directives_missing_colon(file, lines):
+    """Search for directive wrongly typed as comments, like: .. versionchanged 3.6.
+
+    Instead of:
+
+    .. versionchanged:: 3.6
+    """
+    for lno, line in enumerate(lines, start=1):
+        if seems_directive_re.search(line):
+            yield lno, "comment seems to be intended as a directive"
+
+
+@checker(".rst")
+def check_roles_glued_to_the_right(file, lines):
+    r"""Search for roles glued to plurals like: :exc:`Exception`s.
+
+    Instead of:
+       The :exc:`Exceptions`\ s
+
+    The difficulty here is that the following is valid:
+       The :literal:`:exc:`Exceptions``
+    While this is not:
+       The :literal:`:exc:`Exceptions``s
+    """
     for lno, line in enumerate(lines, start=1):
         line = re.sub("``.*?``", "", line).replace("````", "")
         role = re.search(rf"{normal_role}s", line)
         if role:
             yield lno, f"role missing surrogate escape before plural: {role.group(0)!r}"
 
 
-@checker(".rst", severity=2)
-def check_roles(file, lines):
-    """Check for suspicious role constructs."""
+@checker(".rst")
+def check_roles_without_backticks(file, lines):
+    """Search roles without backticks like :func:pdb.main."""
     for lno, line in enumerate(lines, start=1):
         no_backticks = role_with_no_backticks.search(line)
         if no_backticks:
             yield lno, f"role with no backticks: {no_backticks.group(0)!r}"
 
 
-@checker(".rst", severity=2)
-def check_suspicious_backticks_constructs(file, lines):
-    """Check for suspicious constructs with backticks."""
+@checker(".rst")
+def check_roles_preceded_with_backtick(file, lines):
+    """Search for roles preceded by a backtick, like `:fct:`sum`."""
     for lno, line in enumerate(lines, start=1):
         if "`" not in line:
             continue
         if backtick_in_front_of_role.search(line):
             yield lno, "superfluous backtick in front of role"
+
+
+@checker(".rst")
+def check_hyperlinks_missing_space_before_less_than(file, lines):
+    """Search for hyperlinks missing a space, like: `Link text<https://example.com>`."""
```

Comment:
It reads better, but I was trying to group checkers lexicographically by starting them with the name of the syntax they check, all roles-related checkers start with `roles-`, all directive-oriented checkers starts with `directives-`, and all hyperlinks related checkers start with `hyperlinks-`.

I also fear another future checker may check for missing spaces in hyperlinks (but in another place) thus the very very ugly "before less than". Maybe just remove it like you suggest and "uglify" the hypotetical future checker?

Reply:
If the same checker can catch all missing spaces, we can call it `hyperlink-missing-space(s)`

I did some tests (by editing markup on a random page of the devguide, running Sphinx v4.5.0) to see where else spaces can be omitted in hyperlinks (and while I was at it for roles too), and this is the result:
* for links like `` `Python <https://www.python.org/>`_``:
  * `` `Python<https://www.python.org/>`_``:
    * `ERROR: Unknown target name: "python<https://www.python.org/>".`
  * `` `Python < https://www.python.org/>`_``: 
    * `ERROR: Unknown target name: "python < https://www.python.org/>".`
  * `` `Python <https://www.python.org/ >`_``: 
    * `ERROR: Unknown target name: "python <https://www.python.org/ >".`
  * `` ` Python <https://www.python.org/>`_``: 
    * :exclamation: no warnings/errors, renders `` ` Python <https://www.python.org/>`_ ``
  * `` `Python <https://www.python.org/> `_``: 
    * `WARNING: Inline interpreted text or phrase reference start-string without end-string.`
    * renders `` `Python <https://www.python.org/> `_ ``
  * `` ` Python <https://www.python.org/>` _``: 
    * :exclamation: no warnings/errors, renders `Python <https://www.python.org/> _`
* for roles like `` :ref:`Python <test>` `` (after adding a `.. _test:` label):
  * `` :role:`Python<test>` ``:
    * no warnings/errors, renders fine
  * `` :role:`Python < test>` ``:
    * `WARNING: undefined label:  test`, renders `Python`
  * `` :role:`Python <test >` ``:
    * `WARNING: undefined label: test ` [^1], renders `Python`
  * `` :role:` Python <test>` ``: 
    * :exclamation: no warnings/errors, renders `` :ref:` Python <test>` ``
  * `` :role:`Python <test> ` ``:
    * `WARNING: Inline interpreted text or phrase reference start-string without end-string.`
    * renders `` :ref:`Python <test> ` ``

I marked with a :exclamation: the ones that fail silently and should be caught by the linter.  The others are already noticed by Sphinx (at least Sphinx v4.5.0), so they should be caught while building the docs, even though the linter could report them too.  This is also going out of the scope of this issue, so let me know if you want me to create a new issue.

Another solution to group rules could be adding a `category` arg to the decorator for roles, directives, links, etc..  This is a bit overkill if the goal is just to keep those two checkers next to each other, but it's a nice feature that some other linters have.  It will also allow to fix errors in groups using `--enable`/`--disable`.


[^1]: the error should probably use `{label!r}`, since the trailing space is invisible.

Reply:
> If the same checker can catch all missing spaces, we can call it `hyperlink-missing-space(s)`

I don't think we should group checkers: if a single of them report a false positive in a given project, they should be able to disable just this one while keeping the other activated.

> Another solution to group rules could be adding a category (or group) arg to the decorator for roles

Yes it could be nice, but probably as you said overklil (for now) and out of the scope of this PR (already too big).